From 5decf6a25e7e90193e25b6c6a23da8da065e8126 Mon Sep 17 00:00:00 2001
From: Bai Ping <ping.bai@nxp.com>
Date: Tue, 11 Oct 2016 15:43:21 +0800
Subject: [PATCH 1175/5242] MLK-13309-01 ARM: imx: clear the L2_PGE bit on
 imx6sll

commit  e4daf45c615846709ac91b3bed17f233021b949b from
https://source.codeaurora.org/external/imx/linux-imx.git

On i.MX6SLL, the 'L2_PGE' bit in GPC CNTR register is set
by default,this bit must be clear, otherwise, system will
failed to resume from DSM mode if L2 cache is enabled.

Signed-off-by: Bai Ping <ping.bai@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/mach-imx/gpc.c |   10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/arch/arm/mach-imx/gpc.c b/arch/arm/mach-imx/gpc.c
index 78c0a5a..52fe8ab 100644
--- a/arch/arm/mach-imx/gpc.c
+++ b/arch/arm/mach-imx/gpc.c
@@ -20,6 +20,9 @@
 #include "common.h"
 #include "hardware.h"
 
+#define GPC_CNTR		0x000
+#define GPC_CNTR_L2_PGE		22
+
 #define GPC_IMR1		0x008
 #define GPC_PGC_MF_PDN		0x220
 #define GPC_PGC_CPU_PDN		0x2a0
@@ -472,6 +475,13 @@ static int __init imx_gpc_init(struct device_node *node,
 			pr_info("No wakeup source in Mega/Fast domain found!\n");
 	}
 
+	/* clear the L2_PGE bit on i.MX6SLL */
+	if (cpu_is_imx6sll()) {
+		val = readl_relaxed(gpc_base + GPC_CNTR);
+		val &= ~(1 << GPC_CNTR_L2_PGE);
+		writel_relaxed(val, gpc_base + GPC_CNTR);
+	}
+
 	/*
 	 * Clear the OF_POPULATED flag set in of_irq_init so that
 	 * later the GPC power domain driver will not be skipped.
-- 
1.7.9.5

